### PR TITLE
Add strategy for nonlinear extension

### PR DESCRIPTION
**Pull request by nafur**
This PR adds a flexible strategy for the nonlinear extension that replaces the handwritten code in `checkLastCall()`.

### CODE REVIEW

File: src/theory/arith/nl/nonlinear_extension.h [hunk 1]
```diff
@@ -224,6 +225,12 @@ class NonlinearExtension
    */
   void sendLemmas(const std::vector<NlLemma>& out);
 
+  /** Run the strategy for the given effort. */
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.h:228]: Carry over the related documentation from checkLastCall and update it as needed.

File: src/theory/arith/nl/nonlinear_extension.cpp [hunk 2]
```diff
@@ -860,6 +860,85 @@ void NonlinearExtension::presolve()
   Trace("nl-ext") << "NonlinearExtension::presolve" << std::endl;
 }
 
+void NonlinearExtension::runStrategy(Theory::Effort effort,
+                                     const std::vector<Node>& assertions,
+                                     const std::vector<Node>& false_asserts,
+                                     const std::vector<Node>& xts)
+{
+  if (!d_strategy.isStrategyInit())
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.cpp:868]: Nit: we could probably put this in a better place, e.g. a finishInit method called from TheoryArith::finishInit.

**Review by ajreynol — CHANGES_REQUESTED** [review]
This is great.  A few minor requests.  Also, is it possible to already delete the old checkLastCall with this PR?

**Comment by nafur** [comment]
Benchmarks show no differences to master.

File: src/theory/arith/nl/nonlinear_extension.h [hunk 3]
```diff
@@ -148,19 +149,6 @@ class NonlinearExtension
...
-   *
-   * xts : the list of (non-reduced) extended terms in the current context.
-   *
-   * This method adds lemmas to d_im directly.
-   *
-   * If the set lems is non-empty, then no further processing is
-   * necessary. The last call effort check should terminate and these
-   * lemmas should be sent.
-   */
   void checkLastCall(const std::vector<Node>& assertions,
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.h:152]: Remove this declaration of checkLastCall from the header.

**Review by ajreynol — APPROVED** [review]
One minor fix, else LGTM.

File: src/theory/arith/nl/nonlinear_extension.h (hunk 3 above)
Comment (nafur) [reply src/theory/arith/nl/nonlinear_extension.h]: Uh. Hard to believe I forgot to remove that...

**Pull request merged**